ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java

```diff
@@ -60,59 +62,66 @@ public static TopicStreamWriter create(
             serviceContext,
             consumerProperties,
             printTopic),
-        printTopic.getTopic().toString(),
+        printTopic.getTopic(),
         printTopic.getIntervalValue(),
         disconnectCheckInterval,
         printTopic.getLimit());
   }
 
   TopicStreamWriter(
       final SchemaRegistryClient schemaRegistryClient,
-      final KafkaConsumer<String, Bytes> topicConsumer,
+      final KafkaConsumer<Bytes, Bytes> topicConsumer,
       final String topicName,
       final long interval,
       final Duration disconnectCheckInterval,
       final OptionalInt limit
   ) {
-    this.topicConsumer = topicConsumer;
-    this.schemaRegistryClient = schemaRegistryClient;
-    this.topicName = topicName;
+    this.topicConsumer = requireNonNull(topicConsumer, "topicConsumer");
+    this.schemaRegistryClient = requireNonNull(schemaRegistryClient, "schemaRegistryClient");
+    this.topicName = requireNonNull(topicName, "topicName");
     this.interval = interval;
-    this.limit = limit;
-    this.disconnectCheckInterval = Objects
-        .requireNonNull(disconnectCheckInterval, "disconnectCheckInterval");
-
+    this.limit = requireNonNull(limit, "limit");
+    this.disconnectCheckInterval =
+        requireNonNull(disconnectCheckInterval, "disconnectCheckInterval");
     this.messagesWritten = 0;
     this.messagesPolled = 0;
+
+    if (interval < 1) {
+      throw new IllegalArgumentException("INTERVAL must be greater than one, but was: " + interval);
+    }
   }
 
   @Override
   public void write(final OutputStream out) {
     try {
       final RecordFormatter formatter = new RecordFormatter(schemaRegistryClient, topicName);
+
       boolean printFormat = true;
       while (true) {
-        final ConsumerRecords<String, Bytes> records = topicConsumer.poll(disconnectCheckInterval);
+        final ConsumerRecords<Bytes, Bytes> records = topicConsumer.poll(disconnectCheckInterval);
         if (records.isEmpty()) {
-          out.write("\n".getBytes(StandardCharsets.UTF_8));
+          out.write("\n".getBytes(UTF_8));
           out.flush();
-        } else {
-          final List<String> values = formatter.format(records);
-          for (final String value : values) {
-            if (printFormat) {
-              printFormat = false;
-              out.write(("Format:" + formatter.getFormat().name() + "\n")
-                            .getBytes(StandardCharsets.UTF_8));
-            }
-            if (messagesPolled++ % interval == 0) {
-              messagesWritten++;
-              out.write(value.getBytes(StandardCharsets.UTF_8));
-              out.flush();
-            }
+          continue;
+        }
+
+        final List<Supplier<String>> values = formatter.format(records.records(topicName));
+        for (final Supplier<String> value : values) {
+          if (printFormat) {
```

Comment:
is it possible that the format has changed between last time we did a format and now (e.g. first few records happened to deserialize as JSON, but actually it was KAFKA or something like that)? if so should we print it again?

Reply:
That's not how it works.

The formatter previously based its choice of format on the first record and then stuck with that. The new version improves this slightly by looking at the first _batch_ of records. However, once the format is _detected_ its never changed.

This boolean hasn't changed either. It just ensures the format is output before the first row of data.

Again, as discussed offline, this is also changing in the next verison.